keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,980 @@
+CIDR. With this enhancement the user configures a new allocation which uses the
+original cluster CIDR but allocates a /23 instead of a /24 to each node. They
+use the node selector to allocate these IPs only to the nodes with the higher
+capacity.
+
+#### Provision discontiguous ranges
+
+A user wants to create a cluster with 32 nodes each with a capacity to run 256
+pods. This means that each node needs a /24 pod CIDR range and they need a total
+range of /19. However, there aren't enough contiguous IPs in the user's network.
+They can find 4 free ranges of size /21 but no single contiguous /19 range.
+
+Using this enhancement, the user creates 4 different CIDR configurations each
+with a /21 range. The CIDR allocator allocates a /24 range from any of these /21
+ranges to the nodes and the user can now create the cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+This feature does not expand the ability of the NodeIPAM controller to change
+the `Node.Spec.PodCIDRs` field. Once that field is set, either by the controller
+or a third party, it will be treated as immutable. This is particularly relevant
+in situtaitons where users start modifying or deleting the `ClusterCidrConfig`.
+Under no circumstances will the controller attempt to revoke the allocated
+CIDRs (more details on this are discussed below).
+
+### Risks and Mitigations
+
+-   Racing kube-controller-managers. If multiples of the controller are running
+    (as in a HA control plane), how do they coordinate?
+    -   The controllers will coordinate using the existing
+        kube-controller-manager leader election.
+
+## Design Details
+
+### New Resource
+
+```go
+type ClusterCIDRConfig struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec ClusterCIDRConfigSpec
+  Status ClusterCIDRConfigStatus
+}
+
+type ClusterCIDRConfigSpec {
+    // An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    // +required
+    CIDR string
+
+    // This defines which nodes the config is applicable to. A nil selector can
+    // be applied to any node.
+    // +optional
+    NodeSelector *v1.LabelSelector
+
+    // Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    // Users would have to ensure that the kubelet doesn't try to schedule
+    // more pods than are supported by the node's netmask (i.e. the kubelet's
+    // --max-pods flag)
+    // +required
+    PerNodeMaskSize int
+
+    // Is this a good idea?
+    //
+    // If Disabled is set to false, no new PodCIDRs will be allocated from this range
+    // +optional
+    Disabled bool
+}
+
+type ClusterCIDRConfigStatus {
+  Conditions []metav1.Conditions
+}
+
+type ClusterCIDRConfigStatusType string
+type ClusterCIDRConfigStatusReason string
+var (
+  // If the "active" condition is true, this ClusterCIDRConfig can be used by
+  // the controller to allocate PodCIDRs for matching nodes.
+  ClusterCIDRConfigActive ClusterCIDRConfigStatusType = "active"
```

Comment:
Why do we need active/terminating - the finalizer + deletionTimestamp indicates the same thing, no?

Reply:
It's more verbose, for sure. active/terminating by themselves aren't too helpful, I agree. I think they add value when the reasons come into play -- it would help make it clear what's happening if you can see {"active", "cidr-exhausted"} and you know why we're not allocating from that range. Similarly {"terminating", "cidr-in-use"} might explain why the finalizer isn't being cleaned up -- it help users who don't know this Config that well from worrying the controller is broken.

But I don't feel too strongly about this. So if the preferred guidance is to avoid status shenanigans, I'm fine removing it.

Reply:
We can always add API if we need it, but we can't easily take it away.